Centralize Notifications/Tasks

parc-jason opened the pull request:
Signed-off-by: Jason Sherman <jsherman@parcsystems.ca>

Centralize notifications for events/activities, and have a task list (activities that require user response).

Have added in Tasks (user response required) for Connection Invitation (`auto-accept-requests: false`) and Presentation Exchange/Proof Request (`auto-respond-presentation-request: false`).

Have added in Activity Log for the same events. The Activity Log will have the tasks and also their completed state. Role is included in the Activity Log.

Note that this does not change the existing notifications in the application. 

I've added a new notification into the frontend that is for the Notifications screen. These notification indicators are pushed.

However, the Task and Activity Log table data is not pushed, it is queried by the frontend. Note that it uses the existing tables to get the data. Future related work would be: include Credential Offers/Acceptance, possibly store the tasks/events in a new table, replace existing notifications with tasks/activities and push to frontend.

<a href="https://gitpod.io/#https://github.com/hyperledger-labs/business-partner-agent/pull/523"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

domwoe commented:
Sorry, planned to look into it today, but couldn't make it. I'll do on Monday.

etschelp commented on backend/business-partner-agent/src/main/java/org/hyperledger/bpa/impl/ActivitiesManager.java:
```diff
@@ -0,0 +1,258 @@
+import javax.inject.Inject;
+import java.util.*;
+
+public class ActivitiesManager {
```
This class is hard to understand and to read. I guess one part is static config (that should not be static, see comment below) that decides when an event is triggered in the handler. So there is a trigger section. The other part is about mapping the persisted objects back to these events to display in the UI. 

One or all of the following might help to make this class more readable:

1. Add this description as a comment
2. Split this up into a trigger and read section and rename the methods accordingly

What do you think?

parc-jason replied:
Configuration and lists are moved out into `ActivityLogConfig` class, so the `ActivitiesManager` is now just reading the db and returning the results.

domwoe commented on backend/business-partner-agent/src/main/java/org/hyperledger/bpa/controller/api/activity/ActivityRole.java:
```diff
@@ -0,0 +1,39 @@
+    @SerializedName("presentation_exchange_verifier")
+    PRESENTATION_EXCHANGE_VERIFIER;
+
+    ActivityRole() {
```
Just semantics, but we don't send an invitation but a request.  Only when we generate the QR code, we actually create an invitation and the smartphone wallet would send the request.

parc-jason replied:
I changed the constants/enums to `CONNECTION_REQUEST`, it is more concise and clear

domwoe commented on frontend/src/App.vue:
```diff
@@ -126,6 +126,26 @@
             <v-list-item-title>{{ $t("nav.settings") }}</v-list-item-title>
           </v-list-item-content>
         </v-list-item>
+        <v-list-item link :to="{ name: 'Notifications' }">
```
* I'd suggest putting the Notifications menu item as second after Dashboard per default
* We should probably remove the partner/wallet badges now

parc-jason replied:
Moved the Notifications under Dashboard. I removed the Business Partner badge, but left the Wallet badge because we haven't implemented the manual Credential Offer flow yet.

domwoe commented on backend/business-partner-agent/src/main/java/org/hyperledger/bpa/impl/ActivitiesManager.java:
```diff
@@ -0,0 +1,258 @@
+                .build();
+    }
+
+    private String getConnectionAlias(Partner partner) {
```
This comment also applies to the Credential Management

parc-jason replied:
I've returned a `PartnerAPI` object instead of a `connectionAlias` string, then use partnerUtils. Also updated in `CredentialManagement` partner list

domwoe reviewed (commented):
This is an awesome feature!

parc-jason commented on backend/business-partner-agent/src/main/java/org/hyperledger/bpa/controller/api/activity/ActivityRole.java:
```diff
@@ -0,0 +1,39 @@
+import com.google.gson.annotations.SerializedName;
+
+public enum ActivityRole {
+    @JsonProperty("connection_invitation_sender")
```
removed `@SerializedName`

parc-jason commented on backend/business-partner-agent/src/main/java/org/hyperledger/bpa/impl/ActivitiesManager.java:
```diff
@@ -0,0 +1,258 @@
+    }
+
+    static List<ConnectionState> connectionStatesCompleted() {
+        return connectionStates(ConnectionState.ACTIVE, ConnectionState.INVITATION.RESPONSE);
```
I wasn't sure what all the proper states are including how Trinisic/esatus and non BPA agents act.  This is all configurable in `application.yml` now.

parc-jason commented on backend/business-partner-agent/src/main/java/org/hyperledger/bpa/impl/ActivitiesManager.java:
```diff
@@ -0,0 +1,258 @@
+        List<ConnectionState> results = new ArrayList<>();
+        for (ConnectionState state : states)
+            results.add(state);
+        return results;
```
Moved all of this logic in to `application.yml` and `ActivityLogConfig`. But thanks for the tip about the `List.of()`

parc-jason commented on backend/business-partner-agent/src/main/java/org/hyperledger/bpa/impl/ActivitiesManager.java:
```diff
@@ -0,0 +1,258 @@
+                ConnectionState.INVITATION.RESPONSE);
+    }
+
+    static List<ConnectionState> connectionStatesForTasks() {
```
I added a check for whether the flags are set or not. I think this only applies to tasks (no need to check for a to-do if the flag is auto), but it will still show in the ActivityLog.

etschelp commented on backend/business-partner-agent/src/main/resources/application.yml:
```diff
@@ -142,6 +142,18 @@ bpa:
   i18n:
     locale: ${BPA_I18N_LOCALE:en}
     fallbackLocale: ${BPA_I18N_FALLBACK_LOCALE:en}
+  activityLog:
```
Sorry for not being precise enough. What I meant was that whether a task should become active or not should depend on the state of the auto flags set on aca-py. 

I would propose to add a subclass called AcaPyConfig to the RuntimeConfig and initialize it in the [onApplicationEvent](https://github.com/hyperledger-labs/business-partner-agent/blob/9141903fbe47961902141601ab3e2efad271aba7/backend/business-partner-agent/src/main/java/org/hyperledger/bpa/config/RuntimeConfig.java#L90) in there you have access to all the values mainly:

```
   "debug.auto_respond_credential_proposal": true,
    "debug.auto_respond_credential_offer": true,
    "debug.auto_respond_credential_request": true,
    "debug.auto_respond_presentation_proposal": true,
    "debug.auto_respond_presentation_request": true,
    "debug.auto_store_credential": true,
    "debug.auto_verify_presentation": true,
    "debug.auto_accept_invites": true,
    "debug.auto_accept_requests": true,
    "debug.auto_respond_messages": true,
```

If an auto flag is set to false or not set at all then I would expect the task to be active, because otherwise the user will see events and tasks where there are none. Does this make sense?

parc-jason replied:
OK, so you do NOT want to configure which events would be considered COMPLETE/etc, you just want the acapy flags read through RuntimeConfig?

The events that would be for TASK, ACTIVITY LOG and COMPLETED can be statics in Java, and put the debug flags into RuntimeConfig.AcaPyConfig.

parc-jason commented:
I made the changes we discussed this morning: added an AcaPyConfig class to read the debug flags, used statics for the defining what event types were task, activity or completed, removed the extra counts from the dashboard, removed the Update DID from partner vue.

I think before we add new notifications, we should quickly discuss persistence of these activity/tasks... I think it will be quicker and easier to query for the tasks and activity log if we do not read from partner proof or partner/connection tables.

etschelp commented on backend/business-partner-agent/src/main/java/org/hyperledger/bpa/config/RuntimeConfig.java:
```diff
@@ -93,8 +96,40 @@ public void onApplicationEvent(StartupTasks.AcaPyReady event) {
                     .statusConfig()
                     .flatMap(c -> c.getAs("tails_server_base_url", String.class))
                     .isPresent();
+
```
As the AcaPyConfig is a bean as well it makes more sense to have it implement `ApplicationEventListener<StartupTasks.AcaPyReady>`as well and move all this initialization parts into it. 

Also each `ac.statusConfig()` call is a web request. Its more efficient to get the config once and then map it accordingly.

parc-jason replied:
thanks! will clean that up, didn't realize the statusConfig was a new call each time.

etschelp commented on backend/business-partner-agent/src/main/java/org/hyperledger/bpa/config/AcaPyConfig.java:
```diff
@@ -0,0 +1,39 @@
+@NoArgsConstructor
+@Singleton
+public class AcaPyConfig {
+    Boolean autoRespondCredentialOffer;
```
`private` fields are cleaner. Also as no code from the outside should change field values I would also argue that `@Getter` is enough, and you would do the initialization in the event handler. See my comment below.

etschelp commented on backend/business-partner-agent/src/main/java/org/hyperledger/bpa/config/ActivityLogConfig.java:
```diff
@@ -0,0 +1,74 @@
+
+@Singleton
+@NoArgsConstructor
+public class ActivityLogConfig {
```
As this class is used in the event handler and the manager I would inject the AcaPyConfig here.

etschelp commented on backend/business-partner-agent/src/main/java/org/hyperledger/bpa/config/ActivityLogConfig.java:
```diff
@@ -0,0 +1,74 @@
+                ConnectionState.RESPONSE);
+    }
+
+    public List<ConnectionState> getConnectionStatesForTasks() {
```
```java
if (acaPyConfig.getAutoAcceptRequests() != null && acaPyConfig.getAutoAcceptRequests() ) {
    return List.of();
}
return connectionStates(ConnectionState.REQUEST);

```

Like this there is no event and it is also filtered in the manager.

parc-jason replied:
Very nice. Very clean.

etschelp commented on backend/business-partner-agent/src/main/java/org/hyperledger/bpa/config/ActivityLogConfig.java:
```diff
@@ -0,0 +1,74 @@
+                PresentationExchangeState.PRESENTATION_ACKED);
+    }
+
+    public List<PresentationExchangeState> getPresentationExchangeStatesForTasks() {
```
Same as above

etschelp commented on backend/business-partner-agent/src/main/resources/application.yml:
```diff
@@ -142,6 +142,18 @@ bpa:
     fallbackLocale: ${BPA_I18N_FALLBACK_LOCALE:en}
+  activityLog:
+    # use strings instead of arrays so we can override with Environment Variables
+    # these can be comma or " " separated strings. They will be parsed and converted in ActivityLogConfig
```
Is this still needed?

parc-jason replied:
no! good find. removing.

etschelp commented on backend/business-partner-agent/src/main/java/org/hyperledger/bpa/impl/ActivitiesManager.java:
```diff
@@ -0,0 +1,237 @@
+            }
+        }
+        String linkId = p.getId().toString();
+        Optional<PartnerAPI> partner = partnerManager.refreshPartner(p.getPartnerId());
```
Partner refresh is a pretty expensive operation which can take a couple of seconds. I do not see a reason for this. A partner proof must always be linked to a partner, also the partner id will not change. If it does something is completely broken.

I would guess this would work as well:

```java
UUID linkId = p.getPartnerId()
PartnerAPI apiPartner;
Optional<Partner> dbPartner = partnerRepo.findById(linkId);
if (dbPartner.isPresent()) {
    apiPartner = converter.toAPIObject(dbPartner); // org.hyperledger.bpa.impl.util.Converter
} 
```

What we need to do at some point is to make the relation Partner to PartnerProof one to many.

parc-jason replied:
Yes, refresh takes too long, I only did that because the Frontend logic could return different partner names if a partner is returned without all the data. So if the same partner has ALL it's data (VC, profile etc) then it could get a different name than when it returns with only the basic data.  This is a bigger issue that we should file - I think the partner display name logic should be in the backend and stored in the db. So if the alias hasn't been explicitly set, and we find a profile, then we can set the field to that, and when a Verifiable Presentation is found then use that etc (or whatever all the different ways to calculate the name are).  But for now, I'll speed this up using the above and just returning the basic partner.

etschelp replied:
Then we need to reopen https://github.com/hyperledger-labs/business-partner-agent/issues/77 and agree on a design and name resolution order so that it is consistent everywhere. It's probably best to add a partnerName field to the PartnerAPI and move the name resolution code from the frontend into a strategy class in the backend. But anyway calling refresh while resolving activities is not without side effects.

etschelp commented on backend/business-partner-agent/src/main/java/org/hyperledger/bpa/impl/ActivitiesManager.java:
```diff
@@ -0,0 +1,237 @@
+                .role(role)
+                .state(state)
+                .type(type)
+                .partner(partner.get())
```
Get without is present check

etschelp commented on backend/business-partner-agent/src/main/java/org/hyperledger/bpa/impl/ActivitiesManager.java:
```diff
@@ -0,0 +1,237 @@
+                .build();
+    }
+
+    private boolean manualFlag(ActivityType type) {
```
See comment above. I would filter directly use the state list as the filter.

etschelp commented on backend/business-partner-agent/src/main/java/org/hyperledger/bpa/impl/ActivitiesManager.java:
```diff
@@ -0,0 +1,237 @@
+                .role(role)
+                .state(state)
+                .type(type)
+                .partner(partnerManager.refreshPartner(p.getId()).get())
```
See comment below. refresh takes a long long time to complete, when all you need is to load the partner and convert it to the api model.
